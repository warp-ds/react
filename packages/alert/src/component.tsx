import { classNames } from "@chbphone55/classnames";
import React, { PropsWithChildren, ReactElement } from "react";
import { alert as ccAlert } from "@warp-ds/css/component-classes";
import { AlertProps } from "./props.js";
import { ExpandTransition } from "../../_helpers/index.js";
import IconError16 from '@warp-ds/icons/react/error-16';
import IconInfo16 from '@warp-ds/icons/react/info-16';
import IconSuccess16 from '@warp-ds/icons/react/success-16';
import IconWarning16 from '@warp-ds/icons/react/warning-16';

export function Alert({
  show,
  type,
  role = "alert",
  children,
  ...props
}: PropsWithChildren<AlertProps>) {
  const variantClasses = ccAlert[type];
  const iconVariantClasses = ccAlert[`${type}Icon`];

  return (
    <ExpandTransition show={show}>
      <div
        role={role}
        className={classNames(
          props.className,
          `${ccAlert.alert} ${variantClasses}`
        )}
        style={props.style}
      >
        <div className={`${ccAlert.icon} ${iconVariantClasses}`}>
          {iconMap[type]}
        </div>
        <div className={ccAlert.textWrapper}>{children}</div>
      </div>
    </ExpandTransition>
  );
}

const iconMap: {
  [key in AlertProps["type"]]: ReactElement;
} = {
  negative: <IconError16 />,
  positive: <IconSuccess16 />,
  warning: <IconWarning16 />,
  info: <IconInfo16 />,
};
