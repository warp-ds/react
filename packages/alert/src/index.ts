export { Alert } from './component.js';
export type { AlertProps } from './props.js';
