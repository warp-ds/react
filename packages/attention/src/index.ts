export { Attention } from './component.js';
