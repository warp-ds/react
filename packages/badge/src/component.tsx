import { classNames } from '@chbphone55/classnames';
import { badge as ccBadge } from '@warp-ds/css/component-classes';
import React from 'react';
import { BadgeProps } from './props.js';

export function Badge(props: BadgeProps) {
  const {
    children,
    as = 'div',
    variant = 'neutral',
    position,
    ...rest
  } = props;

  return React.createElement(
    as,
    {
      ...(rest as Omit<BadgeProps, 'children'> as {}),
      className: classNames(
        ccBadge.base,
        ccBadge[variant],
        {
            [ccBadge.positionBase]: !!position,
            [ccBadge.positionTL]: position === 'top-left',
            [ccBadge.positionTR]: position === 'top-right',
            [ccBadge.positionBR]: position === 'bottom-right',
            [ccBadge.positionBL]: position === 'bottom-left',
        },
        props.className,
      ),
    },
    children,
  );
}
