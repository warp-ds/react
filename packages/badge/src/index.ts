export { Badge } from './component.js';
export type { BadgeProps } from './props.js';
