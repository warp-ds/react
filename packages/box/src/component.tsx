import { classNames } from '@chbphone55/classnames';
import { box as ccBox } from '@warp-ds/css/component-classes';
import React from 'react';
import { BoxProps } from './props.js';

export function Box(props: BoxProps) {
  const {
    children,
    as = 'div',
    bleed,
    clickable,
    neutral,
    bordered,
    info,
    role,
    ...rest
  } = props;

  return React.createElement(
    as,
    {
      ...(rest as Omit<BoxProps, 'children'> as {}),
      role: role ?? 'region',
      className: classNames(
        ccBox.box,
        {
          [ccBox.bleed]: bleed,
          [ccBox.info]: info,
          [ccBox.neutral]: neutral,
          [ccBox.bordered]: bordered,
          [ccBox.infoClickable]:clickable && info,
          [ccBox.neutralClickable]: clickable && neutral,
          [ccBox.borderedClickable]: clickable && bordered,
        },
        props.className,
      ),
    },
    children,
  );
}
