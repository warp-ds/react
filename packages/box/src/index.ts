export { Box } from './component.js';
export type { BoxProps } from './props.js';
