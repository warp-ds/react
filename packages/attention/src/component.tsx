import React, { forwardRef, useEffect, useMemo, useRef, useState } from 'react';

import { classNames } from '@chbphone55/classnames';
import { i18n } from '@lingui/core';
import { arrowDirectionClassname, opposites, useRecompute as recompute } from '@warp-ds/core/attention';
import { attention as ccAttention } from '@warp-ds/css/component-classes';
import IconClose16 from '@warp-ds/icons/react/close-16';

import { activeAttentionType, getVariant, pointingAtDirection, useAutoUpdatePosition } from '../../_helpers/attention.js';
import { activateI18n } from '../../i18n.js';

import { messages as enMessages } from './locales/en/messages.mjs';
import { messages as fiMessages } from './locales/fi/messages.mjs';
import { messages as nbMessages } from './locales/nb/messages.mjs';
import type { ArrowProps, AttentionProps, ReferenceElement } from './props.js';

const variantClasses = {
  callout: {
    wrapper: ccAttention.callout,
    arrow: ccAttention.arrowCallout,
  },
  highlight: {
    wrapper: ccAttention.highlight,
    arrow: ccAttention.arrowHighlight,
  },
  tooltip: {
    wrapper: ccAttention.tooltip,
    arrow: ccAttention.arrowTooltip,
  },
  popover: {
    wrapper: ccAttention.popover,
    arrow: ccAttention.arrowPopover,
  },
};

export function Attention(props: AttentionProps) {
  let {
    noArrow,
    isShowing,
    children,
    role,
    'aria-label': ariaLabel,
    placement = 'bottom',
    targetEl: initialTargetEl,
    className,
    canClose,
    onDismiss,
    distance = 8,
    skidding = 0,
    flip = false,
    crossAxis = false,
    fallbackPlacements,
    ...rest
  } = props;

  activateI18n(enMessages, nbMessages, fiMessages);

  const wrapperClasses = classNames(ccAttention.base, variantClasses[getVariant(rest, variantClasses)].wrapper);

  const [actualDirection, setActualDirection] = useState(placement);
  // Don't show attention element before its position is computed on first render
  const [isVisible, setIsVisible] = useState<boolean | undefined>(false);

  const isMounted = useRef(true);
  const attentionEl = useRef<HTMLDivElement | null>(null);
  const arrowEl = useRef<HTMLDivElement | null>(null);
  const targetElRef = useRef<ReferenceElement | null>(initialTargetEl?.current || null);

  const attentionState = useMemo(
    () => ({
      get isShowing() {
        return isShowing;
      },
      get isCallout() {
        return rest.callout;
      },
      get actualDirection() {
        return actualDirection;
      },
      set actualDirection(v) {
        setActualDirection(v);
      },
      get directionName() {
        return placement;
      },
      get arrowEl() {
        return arrowEl.current;
      },
      get attentionEl() {
        return attentionEl.current;
      },
      set attentionEl(v) {
        attentionEl.current = v;
      },
      get targetEl() {
        return targetElRef?.current;
      },
      get noArrow() {
        return props.noArrow;
      },
      get distance() {
        return distance;
      },
      get skidding() {
        return skidding;
      },
      get flip() {
        return flip;
      },
      get crossAxis() {
        return crossAxis;
      },
      get fallbackPlacements() {
        return fallbackPlacements;
      },
    }),
    [
      isShowing,
      rest.callout,
      actualDirection,
      placement,
      arrowEl,
      attentionEl,
      targetElRef,
      props.noArrow,
      distance,
      skidding,
      flip,
      crossAxis,
      fallbackPlacements,
    ],
  );

  const defaultAriaLabel = () => `${activeAttentionType(props)} ${!props.noArrow ? pointingAtDirection(actualDirection) : ''}`;

  useEffect(() => {
    // targetEl can be undefined if props.callout is true.
    // However, useAutoUpdatePosition hook is using @warp-ds/core, which uses Floating-ui's computePosition(). Floating-ui's computePosition() requires a defined targetEl to be able to compute the attentionEl's position and the attentionEl's arrow position.
    // When props.callout is true, we only need computePosition() to calculate the callout's arrow position. So, we create a default targetEl for callout that we can pass to the useAutoUpdatePosition hook, in order to avoid Floating-ui from throwing an error.
    if (props.callout && initialTargetEl === undefined) {
      targetElRef.current = document?.createElement('div');
    } else {
      targetElRef.current = initialTargetEl?.current || null;
    }
  }, [props.callout, initialTargetEl]);

  useEffect(() => {
    recompute(attentionState);
  }, [attentionState]);

  useEffect(() => {
    if (isMounted.current) {
      isMounted.current = false;
      // update attention's visibility after first render if showing by default or it's of type callout
      if (isShowing || props.callout) {
        setIsVisible(isShowing);
      }
    } else {
      setIsVisible(isShowing);
    }
  }, [isShowing, props.callout]);

  useAutoUpdatePosition(targetElRef, isShowing, attentionEl, attentionState);

  return !props.callout && initialTargetEl === undefined ? null : (
    <div
      data-testid="attention-el"
      className={classNames(
        {
          [ccAttention.notCallout]: !props.callout,
          invisible: !isVisible && !props.callout,
          hidden: !isVisible && !props.tooltip,
        },
        className,
      )}
      ref={attentionEl}>
      <div
        role={props.role === '' ? undefined : props.tooltip ? 'tooltip' : 'img'}
        aria-label={ariaLabel === '' ? undefined : ariaLabel ?? defaultAriaLabel()}
        className={wrapperClasses}
        id={props.id}>
        {!props.noArrow && <Arrow {...props} ref={arrowEl} direction={actualDirection} />}
        <div className={ccAttention.content}>{props.children}</div>
        {canClose && (
          <button
            type="button"
            aria-label={i18n._(
              /*i18n*/ {
                id: 'attention.aria.close',
                message: 'Close',
                comment: 'Aria label for the close button in attention',
              },
            )}
            onClick={onDismiss}
            onKeyDown={(event) => {
              if (!props.onDismiss) return;
              if (event.key === 'Escape') {
                props.onDismiss();
              }
            }}
            className={ccAttention.closeBtn}>
            <IconClose16 />
          </button>
        )}
      </div>
    </div>
  );
}

const Arrow = forwardRef<HTMLDivElement, ArrowProps>(({ direction, ...rest }, ref) => {
  const arrowDirection = opposites[direction];
  const arrowClasses = classNames(
    ccAttention.arrowBase,
    ccAttention[`arrowDirection${arrowDirectionClassname(arrowDirection)}`],
    variantClasses[getVariant(rest, variantClasses)].arrow,
  );

  return <div data-testid="attention-arrow-el" ref={ref} className={arrowClasses} />;
});
