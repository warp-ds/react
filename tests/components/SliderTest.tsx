import React from 'react';

import { render, fireEvent, screen } from '@testing-library/react';

import { Slider } from '../../packages/slider/src/component';

describe('Slider', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly', () => {
    const { container } = render(<Slider min={0} max={100} />);
    expect(container).toMatchSnapshot();
  });

  it('calls onChange when value changes', async () => {
    const onChange = vi.fn();
    render(<Slider min={0} max={100} onChange={onChange} />);
    const thumb = screen.getByRole('slider');
    await fireEvent.mouseDown(thumb, { clientX: 50 });
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it('calls onKeyDown when thumb is pressed', async () => {
    const onChange = vi.fn();
    render(<Slider min={0} max={100} onChange={onChange} />);
    const thumb = screen.getByRole('slider');
    await fireEvent.keyDown(thumb, { key: 'ArrowRight' });
    // not sure why it gets called twice here
    expect(onChange).toHaveBeenCalledTimes(2);
  });
});
