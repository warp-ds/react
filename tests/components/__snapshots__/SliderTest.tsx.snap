// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Slider > renders correctly 1`] = `
<div>
  <div
    class="touch-pan-y relative w-full h-44 py-2"
  >
    <div
      class="absolute s-bg-disabled-subtle h-4 top-20 rounded-4 w-full  "
    />
    <div
      class="absolute s-bg-primary h-6 top-[19px] rounded-4 "
      style="left: 0px;"
    />
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      class="absolute transition-shadow w-24 h-24 bottom-10 rounded-4 outline-none  border-2 shadow-[--w-shadow-slider] cursor-pointer s-bg-primary s-border-primary hover:s-bg-primary-hover hover:s-border-primary-hover hover:shadow-[--w-shadow-slider-handle-hover] active:s-bg-primary-active active:s-border-primary-active active:shadow-[--w-shadow-slider-handle-active] focus:shadow-[--w-shadow-slider-handle-hover] focus:s-border-primary-hover focus:s-bg-primary-hover"
      role="slider"
      style="transform: translateX(NaNpx);"
      tabindex="0"
    />
  </div>
</div>
`;
